subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/WatchableHierarchies.java

```diff
@@ -138,16 +142,34 @@ private SnapshotHierarchy invalidateUnsupportedFileSystems(SnapshotHierarchy roo
         return recentlyUsedHierarchies;
     }
 
-    private void checkThatNothingExistsInNewWatchableHierarchy(String watchableHierarchy, SnapshotHierarchy vfsRoot) {
-        vfsRoot.visitSnapshotRoots(watchableHierarchy, snapshotRoot -> {
+    @CheckReturnValue
+    private SnapshotHierarchy removeUnwatchedContentAtNewWatchableHierarchy(String unwatchedLocation, SnapshotHierarchy vfsRoot) {
+        RemoveUnwatchedContentVisitor removeUnwatchedContentVisitor = new RemoveUnwatchedContentVisitor(vfsRoot);
+        vfsRoot.visitSnapshotRoots(unwatchedLocation, removeUnwatchedContentVisitor);
+        SnapshotHierarchy newRoot = removeUnwatchedContentVisitor.getVfsRoot();
+        if (newRoot != vfsRoot) {
+            LOGGER.info("Removed existing content under new watchable hierarchy {}", unwatchedLocation);
+        }
+        return newRoot;
+    }
+
+    private class RemoveUnwatchedContentVisitor implements SnapshotHierarchy.SnapshotVisitor {
+        private SnapshotHierarchy vfsRoot;
+
+        public RemoveUnwatchedContentVisitor(SnapshotHierarchy vfsRoot) {
+            this.vfsRoot = vfsRoot;
+        }
+
+        @Override
+        public void visitSnapshotRoot(FileSystemLocationSnapshot snapshotRoot) {
             if (!isInWatchableHierarchy(snapshotRoot.getAbsolutePath()) && !ignoredForWatching(snapshotRoot)) {
-                throw new IllegalStateException(String.format(
-                    "Found existing snapshot at '%s' for unwatched hierarchy '%s'",
-                    snapshotRoot.getAbsolutePath(),
-                    watchableHierarchy
-                ));
```

Comment:
Should we log something here, maybe at `DEBUG`?

Reply:
Added.